fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/visitor/FilterVisitor.java

```diff
@@ -0,0 +1,116 @@
+import com.apple.foundationdb.record.query.expressions.FieldWithComparison;
+import com.apple.foundationdb.record.query.expressions.QueryComponent;
+import com.apple.foundationdb.record.query.expressions.QueryKeyExpressionWithComparison;
+import com.apple.foundationdb.record.query.plan.PlannableIndexTypes;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryFetchFromPartialRecordPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryFilterPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A substitution visitor that pushes a filter below a record fetch if all of the (non-repeated) field are available
+ * in a covering scan.
+ */
+public class FilterVisitor extends RecordQueryPlannerSubstitutionVisitor {
+    public FilterVisitor(@Nonnull final RecordMetaData recordMetadata, @Nonnull final PlannableIndexTypes indexTypes, @Nullable final KeyExpression commonPrimaryKey) {
+        super(recordMetadata, indexTypes, commonPrimaryKey);
+    }
+
+    @Nonnull
+    @Override
+    public RecordQueryPlan postVisit(@Nonnull final RecordQueryPlan recordQueryPlan) {
+        if (recordQueryPlan instanceof RecordQueryFilterPlan) {
+            final RecordQueryFilterPlan filterPlan = (RecordQueryFilterPlan)recordQueryPlan;
+            final QueryComponent filter = filterPlan.getFilter();
+            final Set<KeyExpression> filterFields = new HashSet<>();
+            boolean canCollectFilterFields = findFilterCoveredFields(filter, filterFields);
+
+            if (canCollectFilterFields) {
+                @Nullable RecordQueryPlan newPlan = removeIndexFetch(filterPlan.getChild(), filterFields);
+                if (newPlan != null) {
+                    return new RecordQueryFetchFromPartialRecordPlan(new RecordQueryFilterPlan(newPlan, filter));
+                }
+            }
+        }
+        return recordQueryPlan;
+    }
+
+    // Find equivalent key expressions for fields used by the given filter.
+    // Does not attempt to deal with OneOfThemWithComponent, as the repeated nested field will be spread across multiple
+    // index entries. Reconstituting that as a singleton in a partial record might work for the simplest case, but
+    // could not for multiple such filter conditions.
+    private static boolean findFilterCoveredFields(@Nonnull QueryComponent filter, @Nonnull Set<KeyExpression> filterFields) {
+        if (filter instanceof FieldWithComparison) {
+            filterFields.add(Key.Expressions.field(((FieldWithComparison)filter).getFieldName()));
+            return true;
+        }
+        if (filter instanceof AndOrComponent) {
+            for (QueryComponent child : ((AndOrComponent)filter).getChildren()) {
+                if (!findFilterCoveredFields(child, filterFields)) {
```

Comment:
It can't quite be a property visitor because the property visitors no longer visit `QueryComponent`s (and probably never will, because the new planner uses `QueryPredicate` instead). It's the same concept, but I think we should leave anything like this for a future refactor.